synapse/lib/remcycle.py

```diff
@@ -0,0 +1,876 @@
+          values set by the api configuration method noted below should
+          be encloused with double curly brackets.
+
+    The following configuration values are optional.
+
+        * api: This is a list of two objects, one listing required API values
+          and the second listing default API values.
+            - The first object should be a list of required values.  These
+              must be provided by the user when they call build_http_request.
+            - The second object should be a dictionary of optional values and
+              their defaults.  A user may provide alternative values when
+              calling build_http_request, but sensible defaults should be
+              provided here.
+        * http: A dictionary of key/value items which can provide per-api
+          specific arguements for the creation of HTTPRequest objects.
+        * ingests: A dictionary of Synapse ingest definitions which will be
+          used to create Ingest objects.  During registration of a Nyx object
+          with Hypnos, these will be registered into the Hypnos cortex.
+          Multiple named ingests may be made available for a single API.
+        * vars: This is a list of 2 value items which are stamped into the url
+          value during the construction of the Nyx object.
+
+    See a complete example below:
+
+    ::
+
+        {
+          "api": [
+            [
+              "someplace"
+            ],
+            {
+              "domore": 0
+            }
+          ],
+          "doc": "api example",
+          "http": {
+            "headers": {
+              "token-goodness": "sekrit token"
+            }
+          },
+          "ingests": {
+            "ingest_definition": {
+              "ingest": {
+                "forms": [
+                  [
+                    "inet:ipv4",
+                    {
+                      "var": "ip"
+                    }
+                  ]
+                ],
+                "vars": [
+                  [
+                    "ip",
+                    {
+                      "path": "ip"
+                    }
+                  ]
+                ]
+              }
+            }
+          },
+          "url": "http://vertex.link/api/v4/geoloc/{{someplace}}/info?domore={{domore}}&apikey={APIKEY}",
+          "vars": [
+            ["APIKEY", "8675309"]
+          ]
+        }
+
+    ::
+
+    This example should be interpreted as the following:
+
+        1. The APIKEY value in the 'vars' will be set in the URL, resulting in
+           the following default url:
+
+        ::
+
+            "http://vertex.link/api/v4/geoloc/{someplace}/info?domore={domore}&apikey=8675309"
+
+        ::
+
+        2. The HTTP request will have the header "token-goodness" set to
+           "sekrit token" for the request.
+        3. The caller must provide the "someplace" value in the api_args when
+           calling build_http_request.  The caller may provide the "domore"
```

Comment:
Not sure i want to get into the level of introspection it would take to actually remove template paramter ( ie. the foo= ) part when a {{valu}} is unspecified.  For now, if we come across APIs that would require the removal of the parameter, we'll probably just want to make 2 API defs.